api/alarms/index.js

```diff
@@ -0,0 +1,182 @@
+    }
+    if (log) {
+      this.logElement = log;
+    }
+
+    this.logMessage('Manager: initializing demo');
+
+    this.displayElement.addEventListener('click', (event) => this.handleCancelAlarm(event));
+    chrome.alarms.onAlarm.addListener(alarm => this.handleAlarm(alarm));
+  }
+
+  logMessage(message) {
+    let date = new Date();
+    let pad = (val, len = 2) => val.toString().padStart(len, '0');
+    let h = pad(date.getHours());
+    let m = pad(date.getMinutes());
+    let s = pad(date.getSeconds());
+    let ms = pad(date.getMilliseconds(), 3);
+    let time = `${h}:${m}:${s}.${ms}`;
+
+    let logLine = document.createElement('div');
+    logLine.innerText = `[${time}] ${message}`;
+
+    // Log events in reverse chronological order
+    this.logElement.insertBefore(logLine, this.logElement.firstChild);
+
+    // Log events in chronological order
+    // this.logElement.appendChild(logLine);
+  }
+
+  async handleAlarm(alarm) {
+    let json = JSON.stringify(alarm, null, 2).replace(/\s+/g, ' ');
+    this.logMessage(`Alarm "${alarm.name}" fired\n${json}}`);
+    await this.refreshDisplay();
+  }
+
+  async handleCancelAlarm(event) {
+    if (!event.target.classList.contains('alarm-row__cancel-button')) {
+      return;
+    }
+
+    let name = event.target.parentElement.dataset.name;
+    await this.cancelAlarm(name);
+    await this.refreshDisplay();
+  }
+
+  async cancelAlarm(name) {
+    // TODO: Remove custom promise wrapper once the Alarms API supports promises
+    return new Promise((resolve, reject) => {
+      chrome.alarms.clear(name, (wasCleared) => {
+        if (wasCleared) {
+          this.logMessage(`Manager: canceled alarm "${name}"`)
+          resolve(wasCleared);
+        } else {
+          this.logMessage(`Manager: could not canceled alarm "${name}"`)
+          reject(wasCleared);
+        }
+      });
+    });
+  }
+
+  // Thin wrapper around alarms.create to log creation event
+  createAlarm(name, alarmInfo) {
+    chrome.alarms.create(name, alarmInfo);
+    let json = JSON.stringify(alarmInfo, null, 2).replace(/\s+/g, ' ');
+    this.logMessage(`Created "${name}"\n${json}`);
+    this.refreshDisplay();
+  }
+
+  renderAlarm(alarm, isLast) {
+    let alarmEl = document.createElement('div');
+    alarmEl.classList.add('alarm-row');
+    alarmEl.dataset.name = alarm.name;
+    alarmEl.innerText = JSON.stringify(alarm, 0, 2) + (isLast ? '' : ',');
+
+    let cancelButton = document.createElement('button');
+    cancelButton.classList.add('alarm-row__cancel-button');
+    cancelButton.innerText = 'cancel';
+    alarmEl.appendChild(cancelButton);
+
+    this.displayElement.appendChild(alarmEl);
+  }
+
+  async cancelAllAlarms() {
+    // TODO: Remove custom promise wrapper once the Alarms API supports promises
+    return new Promise((resolve, reject) => {
+      chrome.alarms.clearAll((wasCleared) => {
+        if (wasCleared) {
+          resolve(wasCleared);
+        } else {
+          reject(wasCleared);
```

Comment:
I get that for a demo an error here is useful, but I don't think the Promisified version would treat this as an error—it would simply resolve with false. If you felt like fixing that here and then throwing an error above—e.g.,

```js
cancelAllAlarms().then((result) => {
  if (!result) {
    throw new Error(`could not cancel alarms!`);
  }
})
```

—that might be nicer, but not required.

Reply:
Sorry, I don't follow this snippet. Are you suggesting that we throw in the event handlers declared on lines 11 and 12 or is this supposed to chain off of `chrome.alarms.clearAll`?

Reply:
I'd change line 18 to this:

```js
clearButton.addEventListener('click', async () => {
  const wasCleared = manager.cancelAllAlarms();
  if (!wasCleared) {
    throw new Error('error clearing alarms');
  }
});
```

... but I would also be happy leaving it as-is now. My main point was like, the Promisified-version of the API will eventually resolve with a true or false boolean, which is what you've done now. So that's great.

Reply:
For the sake of reducing churn I'm going to stick with what I have now.